### PR TITLE
feat(evm): Automatically interpret eligible zero gas EVM txs, improving UX to account for wallet client edge cases

### PR DESCRIPTION
**Pull request by Unique-Divine**
# Harden EVM Zero-Gas Fee Exemption

This PR makes `always_zero_gas_contracts` behave as a true fee-exemption policy for EVM calls with `value == 0`, even when wallets submit nonzero or wallet-managed fee fields. It preserves the raw signed Ethereum transaction for identity, signature, hash, and RPC debugging while normalizing the derived execution view so eligible calls do not require or deduct native NIBI gas.

Motivation:

- https://github.com/NibiruChain/sai-website/issues/948
- https://github.com/NibiruChain/sai-website/issues/946

## Key Changes

- Classify eligible EVM txs before fee-sensitive validation in ante and message-server paths, then use zero-gas-aware validation that skips only fee-price checks while preserving hash, chain ID, `to`, `value`, gas limit, intrinsic gas, nonce, and execution semantics.
- Normalize derived `core.Message` fee fields for classified zero-gas execution and set `NoBaseFee` from the zero-gas context, while keeping raw signed tx data unchanged for RPC transaction views and debugging.
- Extend JSON-RPC simulation and tracing paths so `eth_call`, `eth_estimateGas`, `debug_traceCall`, and `debug_traceTransaction` use the same zero-fee execution semantics for allowlisted `value == 0` calls.
- Report zero applied fee in receipts for classified zero-gas transactions while preserving raw signed wallet fee fields in transaction views.
- Add focused ante, message-server, gRPC query, and E2E coverage for allowlisted zero-gas contract calls with wallet-managed fee fields from zero-balance senders.
- Document operator verification for zero-gas EVM fee exemption in `x/evm/README.md`.

## Appendix

Validation run locally:

```bash
go test ./x/evm/...
go test ./eth/rpc/...
go test ./x/evm/evmstate -run 'TestEvmState'
go test ./x/evm/evmstate
just test test/zero_gas_erc20.test.ts
```

### CODE REVIEW

File: x/evm/zero_gas.go [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+		ValidateTxDataAmount(txData),
+		ValidateTxDataTo(txData),
+		ValidateTxDataChainID(txData),
+	} {
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
```
[the remark this answers is not in the document]
Comment (Unique-Divine) [reply x/evm/zero_gas.go:123]: I don’t think this is a valid nil-deref path as described. `AsEthereumData()` uses getters and passes nil fee fields through to go-ethereum `NewTx`; geth’s tx `copy()` initializes nil fee `*big.Int` fields to zero rather than dereferencing them. Raw Ethereum type-2 wallet txs also cannot omit these fields in the way described.

I agree there is a useful hardening/regression-test angle for malformed direct `MsgEthereumTx` protobufs, though. Rather than restoring fee-price validation for classified zero-gas txs, I’ll add or consider a focused test proving nil fee fields do not panic and are handled consistently with the zero-gas policy.

File: eth/rpc/rpcapi/call_tx.go [hunk 2]
```diff
@@ -210,6 +210,33 @@ func (b *Backend) SetTxDefaults(args evm.JsonTxArgs) (evm.JsonTxArgs, error) {
...
+		return false
+	}
+
+	res, err := sudo.NewQueryClient(b.clientCtx).QueryZeroGasActors(
+		context.Background(),
+		&sudo.QueryZeroGasActorsRequest{},
+	)
+	if err != nil || res == nil {
+		return false
+	}
```
[the remark this answers is not in the document]
Comment (Unique-Divine) [reply eth/rpc/rpcapi/call_tx.go:227]: I don’t think this applies. `QueryZeroGasActorsResponse.Actors` is a value field, not a pointer:

`Actors ZeroGasActors`

So `res.Actors == nil` is not possible and the suggested code would not compile. If the zero-gas actor list is empty or uninitialized, `AlwaysZeroGasContracts` is just a nil/empty slice, and ranging over it is safe in Go. The existing `err != nil || res == nil` guard is sufficient for this path.

File: eth/rpc/rpcapi/tx_info.go [hunk 3]
```diff
@@ -343,6 +346,37 @@ func (b *Backend) GetTransactionReceipt(hash gethcommon.Hash) (*TransactionRecei
...
+		return false
+	}
+
+	res, err := sudo.NewQueryClient(b.clientCtx).QueryZeroGasActors(
+		rpc.NewContextWithHeight(height),
+		&sudo.QueryZeroGasActorsRequest{},
+	)
+	if err != nil || res == nil {
+		return false
+	}
```
[the remark this answers is not in the document]
Comment (Unique-Divine) [reply eth/rpc/rpcapi/tx_info.go]: I don’t think this applies. `QueryZeroGasActorsResponse.Actors` is a value field, not a pointer:

`Actors ZeroGasActors`

So `res.Actors == nil` is not possible and the suggested code would not compile. If the zero-gas actor list is empty or uninitialized, `AlwaysZeroGasContracts` is just a nil/empty slice, and ranging over it is safe in Go. The existing `err != nil || res == nil` guard is sufficient for this path.

File: x/evm/evmstate/grpc_query.go [hunk 4]
```diff
@@ -747,7 +782,13 @@ func (k Keeper) TraceBlock(
...
 		// tracing, as this is only a query.
 		msg, _ = core.TransactionToMessage(ethTx, signer, evmCfg.BaseFeeWei)
-		traceResult, logIndex, err := k.TraceEthTxMsg(ctx, evmCfg, txConfig, *msg, req.TraceConfig, tracerConfig)
+		traceCtx := ctx
+		if isZeroGas, _, classErr := evm.IsZeroGasMsgEthereumTx(ctx, k.SudoKeeper, tx); classErr == nil && isZeroGas {
+			traceCtx = evm.WithZeroGasMeta(ctx)
+			normalized := evm.NormalizeZeroGasMessage(*msg)
+			msg = &normalized
+		}
+		traceResult, logIndex, err := k.TraceEthTxMsg(traceCtx, evmCfg, txConfig, *msg, req.TraceConfig, tracerConfig)
```
[the remark this answers is not in the document]
Comment (Unique-Divine) [reply x/evm/evmstate/grpc_query.go:791]: I don’t think this is a valid nil-deref path. In this go-ethereum fork, `core.TransactionToMessage` allocates and populates `msg := &Message{...}` before sender recovery, then returns `msg, err`. A signer error does not make `msg` nil.

This trace path intentionally ignores sender recovery errors because block tracing is query-only and the derived message is still useful for replay/tracing. So I don’t think this is a node-crash issue. At most, we could choose to make trace behavior stricter by returning the signer error, but that would be a behavior change rather than a safety fix.

File: eth/rpc/rpcapi/call_tx.go [hunk 5]
```diff
@@ -210,6 +210,33 @@ func (b *Backend) SetTxDefaults(args evm.JsonTxArgs) (evm.JsonTxArgs, error) {
...
+	for _, raw := range res.Actors.AlwaysZeroGasContracts {
+		if !common.IsHexAddress(raw) {
+			continue
+		}
+		if common.HexToAddress(raw) == *args.To {
+			return true
+		}
+	}
+	return false
+}
```
[the remark this answers is not in the document]
Comment (Unique-Divine) [reply eth/rpc/rpcapi/call_tx.go:238]: Agreed this is a reasonable low-severity cleanup. The two call sites intentionally differ in context source (`latest` for defaults/fill, tx height for receipt reporting), but the address/value/list comparison can be factored to reduce drift. I don’t think this blocks the behavior change, but it is a good follow-up refactor.

File: eth/rpc/rpcapi/call_tx.go [hunk 6]
```diff
@@ -85,7 +82,10 @@ func (b *Backend) SendRawTransaction(data hexutil.Bytes) (common.Hash, error) {
 // SetTxDefaults populates tx message with default values in case they are not
 // provided on the args
 func (b *Backend) SetTxDefaults(args evm.JsonTxArgs) (evm.JsonTxArgs, error) {
-	if args.GasPrice != nil && (args.MaxFeePerGas != nil || args.MaxPriorityFeePerGas != nil) {
+	isZeroGas := b.isZeroGasJsonTxArgsLatest(args)
```
[the remark this answers is not in the document]
Comment (Unique-Divine) [reply eth/rpc/rpcapi/call_tx.go:85]: Partially agreed, but the comment overstates the scope. `isZeroGasJsonTxArgsLatest` returns before the gRPC query when `args.To == nil` or `value != 0`, so contract creation and nonzero-value transfers do not pay this cost.

For `to != nil` and `value == 0`, the lookup is intentional because `SetTxDefaults` needs to classify before rejecting mixed or inconsistent wallet fee fields. That said, if this becomes hot we can cache the zero-gas actor set in the RPC backend or factor the lookup behind a small helper. I would treat this as a low-severity optimization, not a correctness blocker.
